### PR TITLE
`superagent`: refine type for `Response#headers`

### PR DESCRIPTION
**Pull request by OliverJAsh**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [<<url here>>](https://ladjs.github.io/superagent/#response-header-fields)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## superagent ([<kbd>unpkg</kbd>](https://unpkg.com/browse/superagent@latest/))
was missing the following properties:
1. Request
1. Response
1. protocols
1. buffer
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1052aaaad8cc1f997b1e4d1829305a44b97530f1
</p>

**Comment by carnesen** [comment]
Could you link to superagent source code to help convince a skeptic that this more narrow type is an accurate one? For example, I could imagine an http library providing null or undefined instead of an empty string if the header didn't have a value which is allowed in http afaik.

**Comment by OliverJAsh** [comment]
Sure, take a look here:

- https://github.com/ladjs/superagent/blob/1c8338b2e0a3b8f604d08acc7f3cbe305be1e571/src/client.js#L324
- https://github.com/ladjs/superagent/blob/1c8338b2e0a3b8f604d08acc7f3cbe305be1e571/src/client.js#L240

**Review by lukaselmer — APPROVED** [review]

File: types/superagent/index.d.ts [hunk 1]
```diff
@@ -97,7 +97,7 @@ declare namespace request {
         get(header: string): string;
         get(header: "Set-Cookie"): string[];
         header: any;
-        headers: any;
```
Comment (lukaselmer) [inline types/superagent/index.d.ts:100]: `header: { [index: string]: string };`

**Comment by lukaselmer** [comment]
From what I can see here

https://github.com/ladjs/superagent/blob/master/src/client.js#L325

the `.header` field should be typed the same as the `.headers` field. I suggest to also change this

**Review by RyanCavanaugh — APPROVED** [review]

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by OliverJAsh** [comment]
Ready to merge

**Comment by voxpelli** [comment]
Somehow this change caused a regression in my tests here: https://github.com/voxpelli/node-connect-pg-simple/commit/649888c887b418bb67d9ce638cc05435cc1ede74

Somehow the actual value in `res.header['set-cookie']` is a `string[]` rather than a `string`, but only for `set-cookie`, not for the rest

I don't have time to debug more right now, thought I would mention it here though as maybe this is the cause, but could also be some other module in my project that messes things up

**Comment by thetutlage** [comment]
As @voxpelli mentioned, this change has caused regression. A better fix will be to update the following types on the `Response` interface.

```ts
interface Response {
  // ...rest of properties

  get(header: "Set-Cookie" | "set-cookie"): string[];
  header: {
    "set-cookie": string[],
    "Set-Cookie": string[],
    [index: string]: string
  };
  headers: {
    "set-cookie": string[],
    "Set-Cookie": string[],
    [index: string]: string
  };
}
```

**Comment by voxpelli** [comment]
@thetutlage Are you able to make a new PR with that? Referencing this PR? And ideally referencing the piece of code that makes `set-cookie` a special case?

Then this would be a quick fix I think 🙏

**Pull request merged**